### PR TITLE
6313 Кадников Николай

### PR DESCRIPTION
**Pull request by alxmcs**
Это часть кода дипломного проекта одного из моих студентов. Код написан порядка 3 лет назад.
Думаю, его проблемы достаточно очевидны. В ревью хотелось бы увидеть предложения по декомпозиции кода.

### CODE REVIEW

File: statisticsClassification.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import statistics
+from baseFunctions import get_data_set, define_class
+import numpy as np
+from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score
+from sklearn.ensemble import RandomForestClassifier as classifier
+from sklearn.model_selection import train_test_split
+import pandas
+np.seterr(divide='ignore', invalid='ignore')
```
Comment (Airfix982) [inline statisticsClassification.py:8]: Надо определить точку входа в программу

File: statisticsClassification.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+data_set, files, srs = get_data_set("dataset")
+
+chi_points = []
+umann_points = []
+student_points = []
+labels = []
+middles = []
+all_vec = []
+
+train_len = round(len(data_set))
```
Comment (Airfix982) [inline statisticsClassification.py:19]: ненужная строка

File: statisticsClassification.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+train_split = data_set[0:]
+
+for i in range(len(train_split)):
+    for j in range(i + 1, len(train_split)):
```
Comment (Airfix982) [inline statisticsClassification.py:24]: Я бы вынес в функцию добавление элементов в списки, плюс можно оптимизировать код

File: statisticsClassification.py (hunk 1 above)
Comment (Airfix982) [inline statisticsClassification.py:8]: надо добавить точку входа в программу

File: statisticsClassification.py (hunk 2 above)
Comment (Airfix982) [inline statisticsClassification.py:19]: ненужный код

File: statisticsClassification.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        [a, b, c] = statistics.use_stat_for_spectr(train_split[i], train_split[j])
+        label = define_class(files[i], files[j])
+        labels.append(label)
+        middle = (a + b + c)/3
+        middles.append(middle)
+        chi_points.append(a)
+        umann_points.append(b)
+        student_points.append(c)
+        vec = [a, b, c]
+        all_vec.append(vec)
```
Comment (Airfix982) [inline statisticsClassification.py:34]: можно вынести в функцию добавление элементов в массивы, плюс есть пути для улучшения самих добавлений

File: statisticsClassification.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+df.to_csv('stats.csv')
+
+inside_points_chi = []
+inside_points_student = []
+inside_points_umann = []
+
+outside_points_chi = []
+outside_points_student = []
+outside_points_umann = []
+predicted = []
```
Comment (Airfix982) [inline statisticsClassification.py:46]: лишний код

File: statisticsClassification.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+chi_points = np.array(chi_points)
+chi_points = chi_points.reshape(-1,1)
+umann_points = np.array(umann_points)
+umann_points = chi_points.reshape(-1,1)
+student_points = np.array(student_points)
+student_points = chi_points.reshape(-1,1)
+middles = np.array(middles)
+middles = chi_points.reshape(-1,1)
+all_vec = np.array(all_vec)
+all_vec = chi_points.reshape(-1,1)
```
Comment (Airfix982) [inline statisticsClassification.py:63]: можно вынести в функцию изменение типа и размерности

File: baseFunctions.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+from display import read_file
+
+def get_data_set(dyrectory):
+
+    files_name = os.listdir(dyrectory)
+    y = []
+    files_names =[]
+    srs = []
+    for file in files_name:
+        path = dyrectory + "/" + file
```
Comment (Airfix982) [inline baseFunctions.py:13]: лучше через os.path.join

File: baseFunctions.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        [a, b] = read_file(path)
+        y.append(a)
+        files_names.append(file)
+        srs.append(b)
+
+    return y, files_names, srs
+
+
+def audio_to_mel(y,sr):
+    return librosa.feature.melspectrogram(y=y, sr=sr, n_mels=128,fmax = 8000)
```
Comment (Airfix982) [inline baseFunctions.py:23]: в данном случае ненужный код

File: baseFunctions.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+def define_class(y1, y2):
+    dec = 0
+    if (("kitchen" in y1) and ("kitchen" in y2)) and (("mi" in y1) and ("mi" in y2)): dec = 1
+    if (("kitchen" in y1) and ("kitchen" in y2)) and (("android" in y1) and ("android" in y2)): dec = 1
+    if (("kitchen" in y1) and ("kitchen" in y2)) and (("iphone" in y1) and ("iphone" in y2)): dec = 1
+    if (("room" in y1) and ("room" in y2)) and (("mi" in y1) and ("mi" in y2)): dec = 1
+    if (("room" in y1) and ("room" in y2)) and (("android" in y1) and ("android" in y2)): dec = 1
+    if (("room" in y1) and ("room" in y2)) and (("iphone" in y1) and ("iphone" in y2)): dec = 1
+    if (("electro" in y1) and ("electro" in y2)): dec = 1
+    if (("second_pink" in y1) and ("second_pink" in y2)): dec = 1
```
Comment (Airfix982) [inline baseFunctions.py:35]: проверку на наличие подстроки можно вынести в функцию, вместо одинаковых if'ов сделать несколько переменных-условий, проверять их, выход из функции сразу после первого выполнения одного из условий, ну или после не выполнения ни одного

File: baseFunctions.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    return dec
+
+def spectrogramm(x):
+    f1, t1, s1 = spectrogram(x)
+    return f1, t1, s1
```
Comment (Airfix982) [inline baseFunctions.py:41]: лишний в данном случае код

File: baseFunctions.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    return librosa.feature.melspectrogram(y=y, sr=sr, n_mels=128,fmax = 8000)
+
+
+def define_class(y1, y2):
+    dec = 0
+    if (("kitchen" in y1) and ("kitchen" in y2)) and (("mi" in y1) and ("mi" in y2)): dec = 1
+    if (("kitchen" in y1) and ("kitchen" in y2)) and (("android" in y1) and ("android" in y2)): dec = 1
+    if (("kitchen" in y1) and ("kitchen" in y2)) and (("iphone" in y1) and ("iphone" in y2)): dec = 1
+    if (("room" in y1) and ("room" in y2)) and (("mi" in y1) and ("mi" in y2)): dec = 1
+    if (("room" in y1) and ("room" in y2)) and (("android" in y1) and ("android" in y2)): dec = 1
```
Comment (Airfix982) [inline baseFunctions.py:32]: ф

File: statisticsClassification.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+chi_points = []
+umann_points = []
+student_points = []
+labels = []
+middles = []
+all_vec = []
+
+train_len = round(len(data_set))
+
```
Comment (Airfix982) [inline statisticsClassification.py:20]: лишний код

File: baseFunctions.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+def get_data_set(dyrectory):
+
+    files_name = os.listdir(dyrectory)
+    y = []
+    files_names =[]
+    srs = []
+    for file in files_name:
+        path = dyrectory + "/" + file
+        [a, b] = read_file(path)
```
Comment (Airfix982) [inline baseFunctions.py:14]: лучше [a, b] = read_file(os.path.join(dyrectory, file))

File: statisticsClassification.py
Comment (alxmcs) [inline statisticsClassification.py:1]: Да и вообще в данном модуле никакой декомпозиции на функции/классы нет. Хотелось бы увидеть твои более подробные предложения по рефакторингу этого модуля

File: baseFunctions.py (hunk 9 above)
Comment (Airfix982) [inline baseFunctions.py:35]: Здесь я бы сделал как-то так:
def check_for_substring(sub_str: str, str_1: str, str_2: str) -> bool:
    return ((sub_str in str_1) and (sub_str in str_2))


def define_class(y1, y2):
    condition_1 = check_for_substring("kitchen", y1, y2)
    condition_2 = list(check_for_substring("mi", y1, y2),
                       check_for_substring("android", y1, y2),
                       check_for_substring("iphone", y1, y2))

    if condition_1 and any(condition_2):
        return 1
    
    condition_1 = check_for_substring("room", y1, y2)
    if condition_1 and any(condition_2):
        return 1
    
    condition_1 = check_for_substring("electro", y1, y2)
    if condition_1:
        return 1
    
    condition_1 = check_for_substring("second_pink", y1, y2)
    if condition_1:
        return 1
    
    return 0

у нас тут не будет миллиона проверок каждый раз, и все дублирование кода превращено в функцию

File: statisticsClassification.py (hunk 4 above)
Comment (Airfix982) [inline statisticsClassification.py:34]: Этот кусок тоже можно вынести в отдельную функцию, примерно так:

def get_parameters(train_split):
    chi_points = []
    umann_points = []
    student_points = []
    labels = []
    middles = []
    all_vec = []

    for i in range( len(train_split) ):
        for j in range( i + 1, len(train_split) ):
            vec = get_vec( train_split[i], train_split[j] )

            chi_points.append(vec[0])
            umann_points.append(vec[1])
            student_points.append(vec[2])
            labels.append(define_class(files[i], files[j]))
            middle = (vec[0] + vec[1] + vec[2]) / 3
            middles.append(middle)
            all_vec.append(vec)
    return chi_points, umann_points, student_points, labels, middles, all_vec



chi_points, umann_points, student_points, labels, middles, all_vec = get_parameters(train_split)

так красивее, плюс если будет еще какая-нибудь тестовая или еще тренировочная последовательность, не придется дублировать код

File: statisticsClassification.py (hunk 6 above)
Comment (Airfix982) [inline statisticsClassification.py:63]: здесь можно придумать что-то такое


def reshape_element(*points) -> tuple:
    ret_points = list()
    for point in points:
        point = np.array(point)
        point = point.reshape(-1, 1)
        ret_points.append(point)
    return tuple(ret_points)



chi_points, umann_points, student_points, middles, all_vec = reshape_element(chi_points, umann_points, student_points, middles, all_vec)

File: statisticsClassification.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+middles = chi_points.reshape(-1,1)
+all_vec = np.array(all_vec)
+all_vec = chi_points.reshape(-1,1)
+
+
+chi_x_train, chi_x_test, chi_y_train, chi_y_test = train_test_split(chi_points, labels, test_size=0.30)
+umann_x_train, umann_x_test, umann_y_train, umann_y_test = train_test_split(umann_points, labels, test_size=0.30)
+stud_x_train, stud_x_test, stud_y_train, stud_y_test = train_test_split(student_points, labels, test_size=0.30)
+mid_x_train, mid_x_test, mid_y_train, mid_y_test = train_test_split(middles, labels, test_size=0.30)
+all_x_train, all_x_test, all_y_train, all_y_test = train_test_split(all_vec, labels, test_size=0.30)
```
Comment (Airfix982) [inline statisticsClassification.py:70]: Декомпозиция этого участка получается примерно так:

```
def spliting(labels, *points, test_size=0):
    ret_arr = []
    for point in points:
        x_tr, x_ts, y_tr, y_ts = train_test_split(point, labels, test_size=test_size)
        ret_arr.append(tuple(x_tr, x_ts, y_tr, y_ts))
    return tuple(ret_arr)
```

`chi, umann, stud, mid, all = spliting(labels, chi_points, umann_points, student_points, middles, all_vec, test_size=0.30)`

Правда код становится сложнее для понимания и отслеживания, но если будет больше, как я понимаю, классов, то такой подход получается лучше

File: statisticsClassification.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+outside_points_chi = []
+outside_points_student = []
+outside_points_umann = []
+predicted = []
+
+chi_model = classifier()
+umann_model = classifier()
+stud_model = classifier()
+mid_model = classifier()
+all_model = classifier()
```
Comment (Airfix982) [inline statisticsClassification.py:52]: Декомпозиция:

```
def create_models(amount):
    ret_arr = []
    for i in range(amount):
        elem = classifier()
        ret_arr.append(elem)
    return tuple(ret_arr)
```

` chi_model, umann_model, stud_model, mid_model, all_model = create_models(5)`

Ну и я бы создавал модели чуть ниже, ближе к тому участку кода, где они начинают использоваться

File: statisticsClassification.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+umann_x_train, umann_x_test, umann_y_train, umann_y_test = train_test_split(umann_points, labels, test_size=0.30)
+stud_x_train, stud_x_test, stud_y_train, stud_y_test = train_test_split(student_points, labels, test_size=0.30)
+mid_x_train, mid_x_test, mid_y_train, mid_y_test = train_test_split(middles, labels, test_size=0.30)
+all_x_train, all_x_test, all_y_train, all_y_test = train_test_split(all_vec, labels, test_size=0.30)
+
+chi_model.fit(chi_x_train, chi_y_train)
+umann_model.fit(umann_x_train, umann_y_train)
+stud_model.fit(stud_x_train, stud_y_train)
+mid_model.fit(mid_x_train, mid_y_train)
+all_model.fit(X=all_x_train, y=all_y_train)
```
Comment (Airfix982) [inline statisticsClassification.py:76]: На основе предыдущих исправлений


```
def fit_models(**models):
    ret_arr = []
    for model, x_y in models.items():
        model.fit(x_y[0], x_y[1])
        ret_arr.append(model)
    return tuple(ret_arr)
```


`chi_model, umann_model, stud_model, mid_model, all_model = fit_models(chi_model = (chi[::2]), umann_model = (umann[::2]), stud_model = (stud[::2]), mid_model = (mid_[::2]), all_model = (all_))`

File: statisticsClassification.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+
+chi_pred = chi_model.predict(chi_x_test)
+stud_pred = stud_model.predict(stud_x_test)
+umann_pred = umann_model.predict(umann_x_test)
+middles_pred = mid_model.predict(mid_x_test)
+all_pred = all_model.predict(all_x_test)
```
Comment (Airfix982) [inline statisticsClassification.py:83]: Декомпозиция:

```
def model_predict(**model_test):
    ret_arr = []
    for model, x_test in model_test.items():
        pred = model.predict(x_test)
        ret_arr.append(pred)
    return ret_arr
```

`predict = model_predict(chi_model = chi[1], stud_model = stud[1], umann_model = umann[1], mid_model = mid_[1], all_model = all_[1])`

File: statisticsClassification.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+report_chi = [accuracy_score(chi_y_test, chi_pred), f1_score(chi_y_test, chi_pred), precision_score(chi_y_test, chi_pred), recall_score(chi_y_test, chi_pred)]
+report_uman = [accuracy_score(umann_y_test, umann_pred), f1_score(umann_y_test, umann_pred), precision_score(umann_y_test, umann_pred), recall_score(umann_y_test, umann_pred)]
+report_stud = [accuracy_score(stud_y_test, stud_pred), f1_score(stud_y_test, stud_pred), precision_score(stud_y_test, stud_pred), recall_score(stud_y_test, stud_pred)]
+report_mid = [accuracy_score(mid_y_test, middles_pred), f1_score(mid_y_test, middles_pred), precision_score(mid_y_test, middles_pred), recall_score(mid_y_test, middles_pred)]
+report_all = [accuracy_score(all_y_test, all_pred), f1_score(all_y_test, all_pred), precision_score(all_y_test, all_pred), recall_score(all_y_test, all_pred)]
```
Comment (Airfix982) [inline statisticsClassification.py:89]: Это будет выглядеть как-то так:

```
def report_model(**models):
    ret_arr = []
    for predict, y_test in models.items():
        ret_arr.append( [accuracy_score(y_test, predict), f1_score(y_test, predict), precision_score(y_test, predict), recall_score(y_test, predict)] )
    return ret_arr
``


`reports = report_model(predict[0] = chi[3], predict[1] = stud[3], predict[2] = umann[3], predict[3] = mid_[3], predict[4] = all_[3])`

File: statisticsClassification.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+print("1", report_chi)
+print("2", report_uman)
+print("3", report_stud)
+print("4", report_mid)
+print("5", report_all)
```
Comment (Airfix982) [inline statisticsClassification.py:95]: После декомпозиции

```
def print_reports(*reports):
    for report in reports:
        print(reports.index(report), report)
```


`    print_reports(reports)`

File: statisticsClassification.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Airfix982) [reply statisticsClassification.py:89]: Оставлять в таком виде можно, если этот код используется один раз и никому не оказывается, а если, например, продукт с таким кодом идет в т.н. продакшн, то лучше его привести к нормальному виду.

**Comment by alxmcs** [comment]
Зачтено
